gsw/geostrophy.py

```diff
@@ -139,48 +141,43 @@ def distance(lon, lat, p=0):
         distance in meters between adjacent points.
 
     """
-    # This uses the algorithm from pycurrents rather than the one
-    # in GSW-Matlab.
+    lon, lat, = np.atleast_2d(lon), np.atleast_2d(lat)
 
-    if lon.ndim != 1 or lat.ndim != 1:
-        raise ValueError('lon, lat must be 1-D; found shapes %s and %s'
-                         % (lon.shape, lat.shape))
-    if lon.shape != lat.shape:
-        raise ValueError('lon, lat must have same 1-D shape; found %s and %s'
-                         % (lon.shape, lat.shape))
-    if p != 0:
-        if np.iterable(p) and p.shape != lon.shape:
-            raise ValueError('lon, non-scalar p must have same 1-D shape;'
-                             ' found %s and %s'
-                             % (lon.shape, lat.shape))
+    if (lon.size == 1) & (lat.size == 1):
+        raise ValueError('more than one point is needed to compute distance')
+    elif lon.ndim != lat.ndim:
+        raise ValueError('lon, lat must have the same dimension')
 
-        p = np.broadcast_to(p, lon.shape)
+    lon, lat, p = np.broadcast_arrays(lon, lat, p, subok=True)
 
-    slm = slice(None, -1)
-    slp = slice(1, None)
+    p_mid = 0.5 * (p[:, 0:-1] + p[:, 0:-1])
+    lat_mid = 0.5 * (lat[:, :-1] + lat[:, 1:])
```

Comment:
@efiring maybe we can be smarter here and simplify these slices and the `np.atleast2d` above.
(I am out of ideas though.)